File: solutions/2191. Sort the Jumbled Numbers/2191.py
# Time complexity: O(sort)
# Space complexity: O(n)
class Solution:
  def sortJumbled(self, mapping: List[int], nums: List[int]) -> List[int]:
    def getMapped(num: int) -> int:
      if num == 0:
        return mapping[0]

      mapped = 0
      multiplier = 1

      while num > 0:
        digit = num % 10
        mapped += mapping[digit] * multiplier
        num //= 10
        multiplier *= 10

      return mapped

    A = [(getMapped(num), num) for num in nums]
    A.sort(key=lambda x: x[0])

    return [num for _, num in A]

File: solutions/2191. Sort the Jumbled Numbers/2191.cpp
// Time complexity: O(sort)
// Space complexity: O(n)
class Solution {
 public:
  vector<int> sortJumbled(
    const vector<int>& mapping,
    const vector<int>& nums
  ) {
    vector<int> ans;
    vector<pair<int, int>> A;

    for (int i = 0; i < nums.size(); ++i)
      A.push_back(make_pair(getMapped(nums[i], mapping), nums[i]));

    ranges::sort(A, [](const pair<int, int>& a, const pair<int, int>& b) {
      return a.first < b.first;
    });

    for (const auto& [_, num] : A)
      ans.push_back(num);

    return ans;
  }

 private:
  int getMapped(int num, const vector<int>& mapping) {
    if (num == 0)
      return mapping[0];

    int mapped = 0;
    int multiplier = 1;

    while (num > 0) {
      int digit = num % 10;
      mapped += mapping[digit] * multiplier;
      num /= 10;
      multiplier *= 10;
    }

    return mapped;
  }
};